keps/sig-apps/4563-eviction-request-api/README.md

```diff
@@ -0,0 +1,1948 @@
+- [Eviction should be able to request eviction asynchronously](https://github.com/kubernetes/kubernetes/issues/66811)
+- [Reasonable defaults with eviction and PodDisruptionBudget](https://github.com/kubernetes/kubernetes/issues/35318)
+- [New FailedEviction PodDisruptionCondition](https://github.com/kubernetes/kubernetes/issues/128815)
+- [Distinguish PDB error separately in eviction API](https://github.com/kubernetes/kubernetes/issues/125500)
+- [Confusing use of TooManyRequests error for eviction](https://github.com/kubernetes/kubernetes/issues/106286)
+
+### Goals
+
+- Introduce new EvictionRequest API (`evictionrequest.coordination.k8s.io`) and eviction request
+  controller.
+- Allow the EvictionRequest API to be extended with a set of interceptors for each pod.
+- Use API-initiated Eviction API of pods that support the eviction request process, only if there
+  are no active interceptors.
+
+### Non-Goals
+
+- Implement eviction request capabilities in Kubernetes workloads (ReplicaSet, Deployment, etc.).
+- Implement eviction request capabilities for autoscaling, de/scheduling.
+- Synchronizing of the eviction request status to the pod status.
+- Introduce the eviction request concept for types other than pods.
+- Synchronize all pod termination mechanisms (see #4 in the [Motivation](#motivation) section), so that they
+  do not terminate pods under NodeMaintenance/EvictionRequest.
+
+## Proposal
+
+We will introduce a new declarative API called EvictionRequest, whose purpose is to coordinate the
+eviction of a pod from a node. It creates a contract between the eviction requester, the pod, and
+the interceptor (described later). The contract is enforced by the eviction request controller, the
+API type and the API admission.
+
+Pods will carry a new set of annotations about which interceptors are involved in their lifecycle.
+This would allow multiple actors to take an action before the pod is terminated. Each interceptor
+has a priority, and only one at a time may progress with the eviction. If there is no interceptor,
+or the last interceptor (lowest priority) has finished without terminating the pod, the eviction
+request controller will attempt to evict the pod using the existing API-initiated eviction.
+
+Multiple requesters can request the eviction of the same pod, and optionally withdraw their request
+in certain scenarios
+([EvictionRequest Cancellation Examples](#evictionrequest-cancellation-examples)).
+
+We can think of EvictionRequest as a managed and safer alternative to eviction.
+
+See some practical use cases for this feature:
+1. Ability to upscale first before terminating the pods with a Deployment: [Deployment Pod Surge Example](#deployment-pod-surge-example)
+   based on the [EvictionRequest Process](#evictionrequest-process).
+2. Ability to upscale first before terminating the pods with HPA: [HorizontalPodAutoscaler Pod Surge Example](#horizontalpodautoscaler-pod-surge-example)
```

Comment:
I guess dragging along the pdb to have minAvailable == HPA's current target is a pretty hacky thus the need for a non pdb blocked signal.

Reply:
Yes, there are a bunch of problems here, user friendliness, atomicity of actions, and responding to any descheduling (e.g. node drain).